assignment-1/pkg/controller/controller.go

```diff
@@ -0,0 +1,206 @@
+		SnapshotQueue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.DefaultControllerRateLimiter(), "myqueue"),
+	}
+	snapshotInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.handleAdd,
+			UpdateFunc: c.handleUpdate,
+		},
+	)
+	return c
+}
+
+func (c *Controller) handleAdd(obj interface{}) {
+	fmt.Println("handling create snapshot")
+	c.SnapshotQueue.Add(obj)
+}
+
+func (c *Controller) handleUpdate(oldObj interface{}, newObj interface{}) {
+	fmt.Println("handling update snapshot")
+	c.SnapshotQueue.Add(newObj)
+}
+
+func (c *Controller) Run(ch chan struct{}) error {
+	if ok := cache.WaitForCacheSync(ch, c.CacheSynced); !ok {
+		fmt.Println("cache was not synced")
+	}
+	go wait.Until(c.worker, time.Second, ch)
+	<-ch
+	return nil
+}
+
+func (c *Controller) worker() {
+	for c.processItems() {
+	}
+}
+
+func (c *Controller) processItems() bool {
+	item, shutdown := c.SnapshotQueue.Get()
+	if shutdown {
+		return false
+	}
+	c.SnapshotQueue.Forget(item)
+	key, err := cache.MetaNamespaceKeyFunc(item)
+	if err != nil {
+		fmt.Println(err.Error())
+		return false
+	}
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		fmt.Println(err.Error())
+		return false
+	}
+	snapshot, err := c.SnapshotLister.Snapshots(namespace).Get(name)
+	if err != nil {
+		fmt.Println(err.Error())
+		return false
+	}
+	if snapshot.Spec.Operation == "backup" {
+		err := c.createBackup(snapshot)
+		if err != nil {
+			fmt.Println(err.Error())
+			return false
+		}
+	} else if snapshot.Spec.Operation == "restore" {
+		err := c.restoreBackup(snapshot)
+		if err != nil {
+			fmt.Println(err.Error())
+			return false
+		}
+	}
+	return true
+}
+
+func (c *Controller) createBackup(snapshot *v1alpha1.Snapshot) error {
+	// PVC doesn't exists, return error
+	pvc, err := c.KubernetesClientSet.CoreV1().PersistentVolumeClaims(
+		snapshot.Namespace).Get(context.Background(), snapshot.Spec.PvcName,
+		metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	// if volumesnapshot exists, don't create
+	_, err = c.ExternalSnapshotClient.SnapshotV1().VolumeSnapshots(snapshot.Namespace).Get(
+		context.Background(), GetSnapshotName(pvc), metav1.GetOptions{})
+	if err == nil {
+		return err
+	}
+	return c.createSnapshot(pvc, snapshot)
+}
+
+func (c *Controller) createSnapshot(pvc *corev1.PersistentVolumeClaim, snapshot *v1alpha1.Snapshot) error {
+	snapshotName := GetSnapshotName(pvc)
+	snapshotClassname := GetSnapshotClassname()
+	snapTemplate := &volumesnapshotv1.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: snapshotName,
+		},
+		Spec: volumesnapshotv1.VolumeSnapshotSpec{
+			VolumeSnapshotClassName: &snapshotClassname,
+			Source: volumesnapshotv1.VolumeSnapshotSource{
+				PersistentVolumeClaimName: &pvc.Name,
+			},
+		},
+	}
+	_, err := c.ExternalSnapshotClient.SnapshotV1().VolumeSnapshots(pvc.Namespace).Create(
+		context.Background(), snapTemplate, metav1.CreateOptions{})
```

Comment:
instead of creating context.Background(), derive it from parent as arg to the function